mlir/lib/Conversion/QIRToMQTRef/QIRToMQTRef.cpp

```diff
@@ -0,0 +1,492 @@
+    ADD_CONVERT_SIMPLE_GATE(VOp)
+    ADD_CONVERT_SIMPLE_GATE(VdgOp)
+    ADD_CONVERT_SIMPLE_GATE(SXOp)
+    ADD_CONVERT_SIMPLE_GATE(SXdgOp)
+    ADD_CONVERT_SIMPLE_GATE(SWAPOp)
+    ADD_CONVERT_SIMPLE_GATE(iSWAPOp)
+    ADD_CONVERT_SIMPLE_GATE(iSWAPdgOp)
+    ADD_CONVERT_SIMPLE_GATE(PeresOp)
+    ADD_CONVERT_SIMPLE_GATE(PeresdgOp)
+    ADD_CONVERT_SIMPLE_GATE(DCXOp)
+    ADD_CONVERT_SIMPLE_GATE(ECROp)
+  }
+
+  /**
+   * @brief Replaces the call operation with a matching rotation gate operation
+   * from the mqtref dialect
+   *
+   * @param op The call operation that is replaced.
+   * @param gateName The name of the gate.
+   * @param operands The operands of the given operation.
+   * @param ctrlQubits The control Qubits of the given operation.
+   * @param rotationCount The number of rotation degrees.
+   * @param rewriter The PatternRewriter to use.
+   */
+  static void convertRotationGate(LLVM::CallOp& op, const StringRef gateName,
+                                  SmallVector<Value>& operands,
+                                  const SmallVector<Value>& ctrlQubits,
+                                  const size_t rotationCount,
+                                  ConversionPatternRewriter& rewriter) {
+    // extract the degrees from the operand list
+    SmallVector<Value> rotationDegrees;
+    rotationDegrees.reserve(rotationCount);
+    rotationDegrees.insert(
+        rotationDegrees.end(),
+        std::make_move_iterator(operands.end() - rotationCount),
+        std::make_move_iterator(operands.end()));
+    operands.resize(operands.size() - rotationCount);
+
+    // match and replace the fitting gate
+    ADD_CONVERT_ROTATION_GATE(POp)
+    ADD_CONVERT_ROTATION_GATE(UOp)
+    ADD_CONVERT_ROTATION_GATE(U2Op)
+    ADD_CONVERT_ROTATION_GATE(RXOp)
+    ADD_CONVERT_ROTATION_GATE(RYOp)
+    ADD_CONVERT_ROTATION_GATE(RZOp)
+    ADD_CONVERT_ROTATION_GATE(RXXOp)
+    ADD_CONVERT_ROTATION_GATE(RYYOp)
+    ADD_CONVERT_ROTATION_GATE(RZZOp)
+    ADD_CONVERT_ROTATION_GATE(RZXOp)
+    ADD_CONVERT_ROTATION_GATE(XXminusYYOp)
+    ADD_CONVERT_ROTATION_GATE(XXplusYYOp)
+  }
+
+  LogicalResult
+  matchAndRewrite(LLVM::CallOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+
+    // get the name of the operation and prepare the return types
+    const auto fnName = op.getCallee();
+    const auto qubitType = ref::QubitType::get(rewriter.getContext());
+    const auto qregType = ref::QubitRegisterType::get(rewriter.getContext());
+    const auto operands = adaptor.getOperands();
+
+    // get the new operands from the operandMap
+    // workaround to avoid unrealized conversion
+    SmallVector<Value> newOperands;
+    newOperands.reserve(operands.size());
+    for (auto const& val : operands) {
+      if (getState().operandMap.contains(val)) {
+        newOperands.emplace_back(getState().operandMap.at(val));
+      } else {
+        newOperands.emplace_back(val);
+      }
+    }
```

Comment:
Could you maybe explain why exactly that workaround is required and what it is actually working around?